### PR TITLE
Add show unread message button

### PR DESCRIPTION
**Pull request by bogdandarie**

### CODE REVIEW

**Comment by saghul** [comment]
Please going forward rebase from master, don't merge.

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins add to whitelist

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

File: react/features/chat/components/web/MessageContainer.js [hunk 1]
```diff
@@ -8,19 +8,26 @@ import AbstractMessageContainer, { type Props }
...
  * @augments AbstractMessageContainer
  */
-export default class MessageContainer extends AbstractMessageContainer<Props> {
-    /**
-     * Whether or not chat has been scrolled to the bottom of the screen. Used
-     * to determine if chat should be scrolled automatically to the bottom when
-     * the {@code ChatInput} resizes.
-     */
-    _isScrolledToBottom: boolean;
+class MessageContainer extends AbstractMessageContainer<Props, State> {
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.js:24]: Why did you move the `export default` to the bottom?

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
+import { makeStyles } from '@material-ui/core';
+import React, { MouseEventHandler } from 'react';
+import { WithTranslation } from 'react-i18next';
+
+import { translate } from '../../../base/i18n/functions';
+import Icon from '../../../base/icons/components/Icon';
+import { IconArrowDown } from '../../../base/icons/svg/index';
+import { withPixelLineHeight } from '../../../base/styles/functions.web';
+export interface INewMessagesButtonProps extends WithTranslation {
+    inputChatHeight: number;
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:10]: Please add comments for each variable

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    onGoToFirstUnreadMessage: MouseEventHandler;
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:11]: The type should be `() => void` not `MouseEventHandler`

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            position: 'absolute',
+            left: 'calc(50% - 72px)',
+            bottom: '15px'
+        },
+
+        newMessagesButton: {
+            display: 'flex',
+            alignItems: 'center',
+            justifyContent: 'space-between',
+            width: '143px',
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:26]: Do we need to have a fixed width? Considering the content will always be the same I think it should be fine without it. Please test this

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            height: '32px',
+            padding: '6px 8px',
+            border: 'none',
+            borderRadius: theme.shape.borderRadius,
+            backgroundColor: theme.palette.section01,
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:31]: This should be `action02`. Please do not use tokens that are marked for removal

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        arrowDownIconContainer: {
+            height: '20px',
+            width: '20px',
+            display: 'flex',
+            alignItems: 'center',
+            justifyContent: 'center',
+
+            '& svg': {
+                fill: theme.palette.uiBackground,
+                transform: 'matrix(-1, 0, 0, 1, 0, 0);'
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:52]: Why is this necessary?

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            }
+        },
+
+        textContainer: {
+            ...withPixelLineHeight(theme.typography.bodyShortRegular),
+            color: theme.palette.field01
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:58]: This should be `text04`

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @param {Function} onGoToFirstUnreadMessage - Function for lifting up onClick event.
+ * @returns {JSX.Element}
+ */
+function NewMessagesButton({ onGoToFirstUnreadMessage, t }: INewMessagesButtonProps): JSX.Element {
+    const styles = useStyles();
+
+    return (
+        <div
+            className = { `${styles.container}` }>
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:73]: You can just pass `styles.container`. No need for string interpolation

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+            <button
+                aria-label = { t('chat.newMessages') }
+                className = { `${styles.newMessagesButton}` }
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:77]: Same as above

File: react/features/chat/reducer.ts [hunk 10]
```diff
@@ -24,7 +24,8 @@ const DEFAULT_STATE = {
     nbUnreadMessages: 0,
     privateMessageRecipient: undefined,
     lobbyMessageRecipient: undefined,
-    isLobbyChatActive: false
+    isLobbyChatActive: false,
+    inputChatHeight: 0
```
Comment (robertpin) [inline react/features/chat/reducer.ts:28]: Not used. Please revert all changes of this file

File: tsconfig.json [hunk 11]
```diff
@@ -16,4 +16,4 @@
     "exclude": [
         "node_modules"
     ]
-}
+}
```
Comment (robertpin) [inline tsconfig.json:19]: Revert change please

**Review by robertpin — CHANGES_REQUESTED** [review]

**Comment by damencho** [comment]
Jenkins add to whitelist

**Comment by damencho** [comment]
Jenkins test this please.

File: react/features/chat/components/web/MessageContainer.js [hunk 12]
```diff
@@ -43,12 +55,17 @@ export default class MessageContainer extends AbstractMessageContainer<Props> {
...
         this._messagesListEndRef = React.createRef();
 
-        this._onChatScroll = this._onChatScroll.bind(this);
+        // Bind event handlers so they are only bound once for every instance.
+        this._createBottomListObserver = this._createBottomListObserver.bind(this);
+        this._handleIntersectBottomList = this._handleIntersectBottomList.bind(this);
+
+        this._findFirstUnreadMessage = this._findFirstUnreadMessage.bind(this);
+        this._isMessageVisible = this._isMessageVisible.bind(this);
+        this._onChatScroll = throttle(this._onChatScroll.bind(this), 2000, { leading: true });
```
[the remark this answers is not in the document]
Comment (bogdandarie) [reply react/features/chat/components/web/MessageContainer.js]: Yes, it is, I have changed it to 300 ms, please let me know if it is ok

File: react/features/chat/components/web/MessageContainer.js [hunk 13]
```diff
@@ -70,43 +87,90 @@ export default class MessageContainer extends AbstractMessageContainer<Props> {
...
-                <div ref = { this._messagesListEndRef } />
+            <div id = 'chat-conversation-container'>
+                <div
+                    aria-labelledby = 'chat-header'
+                    id = 'chatconversation'
+                    onScroll = { this._onChatScroll }
+                    ref = { this._messageListRef }
+                    role = 'log'
+                    tabIndex = { 0 }>
+                    <div />
```
[the remark this answers is not in the document]
Comment (bogdandarie) [reply react/features/chat/components/web/MessageContainer.js]: Is not, I have deleted

File: react/features/chat/components/web/MessageContainer.js [hunk 14]
```diff
@@ -70,43 +87,90 @@ export default class MessageContainer extends AbstractMessageContainer<Props> {
...
+     * Otherwise update hasNewMessages from component state.
+     *
+     * @inheritdoc
+     * * @returns {void}
      */
-    maybeUpdateBottomScroll() {
-        if (this._isScrolledToBottom) {
-            this.scrollToBottom(false);
+    componentDidUpdate(prevProps) {
+        const hasNewMessages = this.props.messages !== prevProps.messages
```
[the remark this answers is not in the document]
Comment (bogdandarie) [reply react/features/chat/components/web/MessageContainer.js]: Yes, length is enough

File: react/features/chat/components/web/MessageContainer.js [hunk 15]
```diff
@@ -70,43 +87,90 @@ export default class MessageContainer extends AbstractMessageContainer<Props> {
...
+            && this.props.messages.length !== prevProps.messages.length;
+
+        if (hasNewMessages && this.state.isScrolledToBottom) {
+            this.scrollToElement(true, null);
```
[the remark this answers is not in the document]
Comment (bogdandarie) [reply react/features/chat/components/web/MessageContainer.js]: Agree, I changed to false

File: react/features/chat/components/web/NewMessagesButton.tsx (hunk 6 above)
Comment (bogdandarie) [reply react/features/chat/components/web/NewMessagesButton.tsx]: I have deleted, I took it from Figma, not really necessary

File: react/features/chat/components/web/ChatInput.tsx [hunk 16]
```diff
@@ -101,7 +93,9 @@ class ChatInput extends Component<Props, State> {
     componentDidMount() {
         if (isMobileBrowser()) {
             // Ensure textarea is not focused when opening chat on mobile browser.
-            this._textArea && this._textArea.blur();
+            if (this._textArea && this._textArea.current) {
```
Comment (robertpin) [inline react/features/chat/components/web/ChatInput.tsx:96]: I think you could've just done `this._textArea?.current && this._textArea.current.blur()`

File: react/features/chat/components/web/ChatInput.tsx [hunk 17]
```diff
@@ -155,7 +149,9 @@ class ChatInput extends Component<Props, State> {
      * @returns {void}
      */
     _focus() {
-        this._textArea && this._textArea.focus();
+        if (this._textArea && this._textArea.current) {
```
Comment (robertpin) [inline react/features/chat/components/web/ChatInput.tsx:152]: Same as above

File: react/features/chat/components/web/MessageContainer.tsx [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+// @ts-ignore
+import ChatMessageGroup from './ChatMessageGroup';
+import NewMessagesButton from './NewMessagesButton';
+
+interface State {
+
+    /**
+     * Weather or not message container has received new messages.
+     */
+
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:20]: Remove unnecessary empty line

File: react/features/chat/components/web/MessageContainer.tsx [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+import AbstractMessageContainer, { type Props } from '../AbstractMessageContainer';
+
+// @ts-ignore
+import ChatMessageGroup from './ChatMessageGroup';
+import NewMessagesButton from './NewMessagesButton';
+
+interface State {
+
+    /**
+     * Weather or not message container has received new messages.
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:18]: Whether*

File: react/features/chat/components/web/MessageContainer.tsx [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+     */
+
+    hasNewMessages: boolean;
+
+    /**
+     * Weather or not scroll position is at the bottom of container.
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:24]: Whether*

File: react/features/chat/components/web/MessageContainer.tsx [hunk 21]
```diff
@@ -0,0 +1,320 @@
...
+     */
+    constructor(props: Props) {
+        super(props);
+
+        this._messageListRef = React.createRef<HTMLDivElement>();
+        this._messagesListEndRef = React.createRef<HTMLDivElement>();
+
+        // Bind event handlers so they are only bound once for every instance.
+        this._handleIntersectBottomList = this._handleIntersectBottomList.bind(this);
+
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:81]: Remove empty line please

File: react/features/chat/components/web/MessageContainer.tsx [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+     * If the user receive a new message scroll automatically to the bottom if scroll position was at the bottom.
+     * Otherwise update hasNewMessages from component state.
+     *
+     * @inheritdoc
+     * @returns {void}
+     */
+    componentDidUpdate(prevProps: Props) {
+        const hasNewMessages = this.props.messages.length !== prevProps.messages.length;
+
+        if (hasNewMessages && this.state.isScrolledToBottom) {
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:152]: Considering both if conditions depend on `hasNewMessages` we could have a parent if that only checks for `hasNewMessages` and inside an if - else

File: react/features/chat/components/web/MessageContainer.tsx [hunk 23]
```diff
@@ -0,0 +1,320 @@
+/* eslint-disable lines-around-comment */
+import throttle from 'lodash/throttle';
+import React, { RefObject } from 'react';
+import { scrollIntoView } from 'seamless-scroll-polyfill';
+
+// @ts-ignore
+import { MESSAGE_TYPE_REMOTE } from '../../constants';
+// @ts-ignore
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:8]: AbstractMessageContainer is now in TS. Remove this comment

File: react/features/chat/components/web/MessageContainer.tsx [hunk 24]
```diff
@@ -0,0 +1,320 @@
...
+import AbstractMessageContainer, { type Props } from '../AbstractMessageContainer';
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:9]: Remove `type`. In TS you can just import `{ Props }`

File: react/features/chat/components/web/MessageContainer.tsx [hunk 25]
```diff
@@ -0,0 +1,320 @@
...
+        const scrollTo = element ? element : this._messagesListEndRef.current;
+        const block = element ? 'center' : 'nearest';
+
+        scrollIntoView(scrollTo as Element, {
+            behavior: withAnimation ? 'smooth' : 'auto',
+            block
+        });
+    }
+
+    _getMessagesGroupedBySender: () => Array<Array<any>>;
```
Comment (robertpin) [inline react/features/chat/components/web/MessageContainer.tsx:192]: This is unnecessary. Please remove it

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 26]
```diff
@@ -0,0 +1,93 @@
+import { makeStyles } from '@material-ui/core';
+import React from 'react';
+import { WithTranslation } from 'react-i18next';
+
+import { translate } from '../../../base/i18n/functions';
+import Icon from '../../../base/icons/components/Icon';
+import { IconArrowDown } from '../../../base/icons/svg/index';
+import { withPixelLineHeight } from '../../../base/styles/functions.web';
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:8]: Please add an empty line after import group

File: react/features/chat/components/web/NewMessagesButton.tsx [hunk 27]
```diff
@@ -0,0 +1,93 @@
...
+ * @param {Function} onGoToFirstUnreadMessage - Function for lifting up onClick event.
+ * @returns {JSX.Element}
+ */
+function NewMessagesButton({ onGoToFirstUnreadMessage, t }: INewMessagesButtonProps): JSX.Element {
+    const styles = useStyles();
+
+    return (
+        <div
+            className = { styles.container }>
+
```
Comment (robertpin) [inline react/features/chat/components/web/NewMessagesButton.tsx:77]: Please remove unnecessary empty line

**Review by robertpin — CHANGES_REQUESTED** [review]

**Review by robertpin — APPROVED** [review]

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by robertpin** [comment]
Jenkins test this please

**Comment by bogdandarie** [comment]
Jenkins test this please

**Comment by saghul** [comment]
@bogdandarie Nice work! Will you be working on the equivalent, for mobile, next?

**Comment by bogdandarie** [comment]
> @bogdandarie Nice work! Will you be working on the equivalent, for mobile, next?

I don't know for sure, until now I have never worked on mobile devices

**Comment by saghul** [comment]
It's React Native, so a close cousin to what you've done already :-) Let's discuss later today.

**Comment by bogdandarie** [comment]
> It's React Native, so a close cousin to what you've done already :-) Let's discuss later today.

Ok, looking forward to discuss :D
